### PR TITLE
Migrate to using Cecil types to resolve generics and types equality

### LINKED ISSUE #9090: CompileAvaloniaXamlTask fails when binding to IDicitionary<TKey, TValue> properties
### Discussed in https://github.com/AvaloniaUI/Avalonia/discussions/8809

<div type='discussions-op-text'>

<sup>Originally posted by **Dryadxon** August 22, 2022</sup>
Hi, before I open an issue for an intended feature, can someone help me to understand what I'm doing wrong?
Recently, I tried setting `x:CompileBindings` to `True` in all my projects, but I stumbled into one which fails to build when the `ComboBox.Items` property is bound to either the `Keys` or `Values` properties of the `IDicitionary<TKey, TValue>` interface in the ViewModel.  
For completeness, I tried other controls with the `Items` property, and the compilation still fails with identical errors.  
I even tried making a copy of the `IDictionary<TKey, TValue>` interface, made a concrete type out of it, which is just a wrapper of `Dictionary<TKey, TValue>`, and to my surprise it did compile just fine.  
I can't see why the build fails if I don't use a `ReflectionBinding` with `IDictionary<TKey, TValue>` properties, looking at [this link](https://docs.avaloniaui.net/docs/data-binding/compiledbindings#known-limitations) everything should just works, can someone hint me why the build fails?

- OS: Linux (Fedora 36)
- Version 11.0.0-preview1, 0.10.18.
- Dotnet Version: 6.0.400

EDIT: I [created](https://github.com/Dryadxon/AvaloniaCompileBindingsIssue) a simple MVVM project to reproduce the issue I found if anyone needs it.

<details>
  <summary>Build log of a repro project</summary>
  
  ```
  /home/admin/.nuget/packages/avalonia/0.10.18/buildTransitive/AvaloniaBuildTasks.targets(98,5): error MSB4018: The "CompileAvaloniaXamlTask" task failed unexpectedly. [/home/admin/AvaloniaDemo/AvaloniaDemo.csproj]
/home/admin/.nuget/packages/avalonia/0.10.18/buildTransitive/AvaloniaBuildTasks.targets(98,5): error MSB4018: System.ArgumentException: Member 'System.Collections.Generic.ICollection`1' is declared in another module and needs to be imported [/home/admin/AvaloniaDemo/AvaloniaDemo.csproj]
/home/admin/.nuget/packages/avalonia/0.10.18/buildTransitive/AvaloniaBuildTasks.targets(98,5): error MSB4018:    at Mono.Cecil.MetadataBuilder.LookupToken(IMetadataTokenProvider provider) [/home/admin/AvaloniaDemo/AvaloniaDemo.csproj]
/home/admin/.nuget/packages/avalonia/0.10.18/buildTransitive/AvaloniaBuildTasks.targets(98,5): error MSB4018:    at Mono.Cecil.SignatureWriter.MakeTypeDefOrRefCodedRID(TypeReference type) [/home/admin/AvaloniaDemo/AvaloniaDemo.csproj]
/home/admin/.nuget/packages/avalonia/0.10.18/buildTransitive/AvaloniaBuildTasks.targets(98,5): error MSB4018:    at Mono.Cecil.SignatureWriter.WriteTypeSignature(TypeReference type) [/home/admin/AvaloniaDemo/AvaloniaDemo.csproj]
/home/admin/.nuget/packages/avalonia/0.10.18/buildTransitive/AvaloniaBuildTasks.targets(98,5): error MSB4018:    at Mono.Cecil.SignatureWriter.WriteMethodSignature(IMethodSignature method) [/home/admin/AvaloniaDemo/AvaloniaDemo.csproj]
/home/admin/.nuget/packages/avalonia/0.10.18/buildTransitive/AvaloniaBuildTasks.targets(98,5): error MSB4018:    at Mono.Cecil.MetadataBuilder.GetMethodSignature(IMethodSignature method) [/home/admin/AvaloniaDemo/AvaloniaDemo.csproj]
/home/admin/.nuget/packages/avalonia/0.10.18/buildTransitive/AvaloniaBuildTasks.targets(98,5): error MSB4018:    at Mono.Cecil.MetadataBuilder.GetMemberRefSignature(MemberReference member) [/home/admin/AvaloniaDemo/AvaloniaDemo.csproj]
/home/admin/.nuget/packages/avalonia/0.10.18/buildTransitive/AvaloniaBuildTasks.targets(98,5): error MSB4018:    at Mono.Cecil.MetadataBuilder.CreateMemberRefRow(MemberReference member) [/home/admin/AvaloniaDemo/AvaloniaDemo.csproj]
/home/admin/.nuget/packages/avalonia/0.10.18/buildTransitive/AvaloniaBuildTasks.targets(98,5): error MSB4018:    at Mono.Cecil.MetadataBuilder.GetMemberRefToken(MemberReference member) [/home/admin/AvaloniaDemo/AvaloniaDemo.csproj]
/home/admin/.nuget/packages/avalonia/0.10.18/buildTransitive/AvaloniaBuildTasks.targets(98,5): error MSB4018:    at Mono.Cecil.MetadataBuilder.LookupToken(IMetadataTokenProvider provider) [/home/admin/AvaloniaDemo/AvaloniaDemo.csproj]
/home/admin/.nuget/packages/avalonia/0.10.18/buildTransitive/AvaloniaBuildTasks.targets(98,5): error MSB4018:    at Mono.Cecil.Cil.CodeWriter.WriteOperand(Instruction instruction) [/home/admin/AvaloniaDemo/AvaloniaDemo.csproj]
/home/admin/.nuget/packages/avalonia/0.10.18/buildTransitive/AvaloniaBuildTasks.targets(98,5): error MSB4018:    at Mono.Cecil.Cil.CodeWriter.WriteInstructions() [/home/admin/AvaloniaDemo/AvaloniaDemo.csproj]
/home/admin/.nuget/packages/avalonia/0.10.18/buildTransitive/AvaloniaBuildTasks.targets(98,5): error MSB4018:    at Mono.Cecil.Cil.CodeWriter.WriteResolvedMethodBody(MethodDefinition method) [/home/admin/AvaloniaDemo/AvaloniaDemo.csproj]
/home/admin/.nuget/packages/avalonia/0.10.18/buildTransitive/AvaloniaBuildTasks.targets(98,5): error MSB4018:    at Mono.Cecil.Cil.CodeWriter.WriteMethodBody(MethodDefinition method) [/home/admin/AvaloniaDemo/AvaloniaDemo.csproj]
/home/admin/.nuget/packages/avalonia/0.10.18/buildTransitive/AvaloniaBuildTasks.targets(98,5): error MSB4018:    at Mono.Cecil.MetadataBuilder.AddMethod(MethodDefinition method) [/home/admin/AvaloniaDemo/AvaloniaDemo.csproj]
/home/admin/.nuget/packages/avalonia/0.10.18/buildTransitive/AvaloniaBuildTasks.targets(98,5): error MSB4018:    at Mono.Cecil.MetadataBuilder.AddMethods(TypeDefinition type) [/home/admin/AvaloniaDemo/AvaloniaDemo.csproj]
/home/admin/.nuget/packages/avalonia/0.10.18/buildTransitive/AvaloniaBuildTasks.targets(98,5): error MSB4018:    at Mono.Cecil.MetadataBuilder.AddType(TypeDefinition type) [/home/admin/AvaloniaDemo/AvaloniaDemo.csproj]
/home/admin/.nuget/packages/avalonia/0.10.18/buildTransitive/AvaloniaBuildTasks.targets(98,5): error MSB4018:    at Mono.Cecil.MetadataBuilder.AddTypes() [/home/admin/AvaloniaDemo/AvaloniaDemo.csproj]
/home/admin/.nuget/packages/avalonia/0.10.18/buildTransitive/AvaloniaBuildTasks.targets(98,5): error MSB4018:    at Mono.Cecil.MetadataBuilder.BuildTypes() [/home/admin/AvaloniaDemo/AvaloniaDemo.csproj]
/home/admin/.nuget/packages/avalonia/0.10.18/buildTransitive/AvaloniaBuildTasks.targets(98,5): error MSB4018:    at Mono.Cecil.MetadataBuilder.BuildModule() [/home/admin/AvaloniaDemo/AvaloniaDemo.csproj]
/home/admin/.nuget/packages/avalonia/0.10.18/buildTransitive/AvaloniaBuildTasks.targets(98,5): error MSB4018:    at Mono.Cecil.MetadataBuilder.BuildMetadata() [/home/admin/AvaloniaDemo/AvaloniaDemo.csproj]
/home/admin/.nuget/packages/avalonia/0.10.18/buildTransitive/AvaloniaBuildTasks.targets(98,5): error MSB4018:    at Mono.Cecil.ModuleWriter.<>c.<BuildMetadata>b__2_0(MetadataBuilder builder, MetadataReader _) [/home/admin/AvaloniaDemo/AvaloniaDemo.csproj]
/home/admin/.nuget/packages/avalonia/0.10.18/buildTransitive/AvaloniaBuildTasks.targets(98,5): error MSB4018:    at Mono.Cecil.ModuleDefinition.Read[TItem,TRet](TItem item, Func`3 read) [/home/admin/AvaloniaDemo/AvaloniaDemo.csproj]
/home/admin/.nuget/packages/avalonia/0.10.18/buildTransitive/AvaloniaBuildTasks.targets(98,5): error MSB4018:    at Mono.Cecil.ModuleWriter.BuildMetadata(ModuleDefinition module, MetadataBuilder metadata) [/home/admin/AvaloniaDemo/AvaloniaDemo.csproj]
/home/admin/.nuget/packages/avalonia/0.10.18/buildTransitive/AvaloniaBuildTasks.targets(98,5): error MSB4018:    at Mono.Cecil.ModuleWriter.Write(ModuleDefinition module, Disposable`1 stream, WriterParameters parameters) [/home/admin/AvaloniaDemo/AvaloniaDemo.csproj]
/home/admin/.nuget/packages/avalonia/0.10.18/buildTransitive/AvaloniaBuildTasks.targets(98,5): error MSB4018:    at Mono.Cecil.ModuleWriter.WriteModule(ModuleDefinition module, Disposable`1 stream, WriterParameters parameters) [/home/admin/AvaloniaDemo/AvaloniaDemo.csproj]
/home/admin/.nuget/packages/avalonia/0.10.18/buildTransitive/AvaloniaBuildTasks.targets(98,5): error MSB4018:    at Mono.Cecil.ModuleDefinition.Write(String fileName, WriterParameters parameters) [/home/admin/AvaloniaDemo/AvaloniaDemo.csproj]
/home/admin/.nuget/packages/avalonia/0.10.18/buildTransitive/AvaloniaBuildTasks.targets(98,5): error MSB4018:    at Mono.Cecil.AssemblyDefinition.Write(String fileName, WriterParameters parameters) [/home/admin/AvaloniaDemo/AvaloniaDemo.csproj]
/home/admin/.nuget/packages/avalonia/0.10.18/buildTransitive/AvaloniaBuildTasks.targets(98,5): error MSB4018:    at Avalonia.Build.Tasks.XamlCompilerTaskExecutor.Compile(IBuildEngine engine, String input, String[] references, String projectDirectory, String output, Boolean verifyIl, MessageImportance logImportance, String strongNameKey, Boolean patchCom, Boolean skipXamlCompilation, Boolean debuggerLaunch) [/home/admin/AvaloniaDemo/AvaloniaDemo.csproj]
/home/admin/.nuget/packages/avalonia/0.10.18/buildTransitive/AvaloniaBuildTasks.targets(98,5): error MSB4018:    at Avalonia.Build.Tasks.CompileAvaloniaXamlTask.Execute() [/home/admin/AvaloniaDemo/AvaloniaDemo.csproj]
/home/admin/.nuget/packages/avalonia/0.10.18/buildTransitive/AvaloniaBuildTasks.targets(98,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() [/home/admin/AvaloniaDemo/AvaloniaDemo.csproj]
/home/admin/.nuget/packages/avalonia/0.10.18/buildTransitive/AvaloniaBuildTasks.targets(98,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask) [/home/admin/AvaloniaDemo/AvaloniaDemo.csproj]
  ```
</details></div>

### LINKED ISSUE #10485: CompileBinding are unable to bind to nested property of generic class.
Using Avalonia `11.0.999-cibuild0030320-beta`.

Set this in `csproj`

    <AvaloniaUseCompiledBindingsByDefault>true</AvaloniaUseCompiledBindingsByDefault>

Add a ComboBox with `SelectedItem` binding.

Run.

> System.BadImageFormatException: An attempt was made to load a program with an incorrect format.

In my project, I was having the error at compile-time, but I reproduced it in the blank template project and this time the error shows up at runtime instead. No idea why. I was also having another error that I'm unable to reproduce, but it could be directly related to this.

I created a repo project here
[Avalonia11-CompileBinding.zip](https://github.com/AvaloniaUI/Avalonia/files/10832993/Avalonia11-CompileBinding.zip)

OS: Garuda Linux

### PR DESCRIPTION
**Pull request by maxkatz6**
cc @kekekeks @MrJul

This is a heavy PR, but these changes do make sense, solving most of our generics related problems.

Essentially, this PR removes old code related to CecilHelpers.TransformGeneric and CecilHelpers.Equals.
And adds copy-pasted internal implementations from the Cecil itself (that never have been made public).
It now includes:
- GenericParameterResolver
- GenericTypeResolver
- MethodReferenceEqualityComparer
- TypeReferenceEqualityComparer

Where GenericTypeResolver always works in a context. In a context of "owner" generic type and "owner" generic method (one of them or both, or none if no generics). Which would require passing GenericTypeResolver down from the root types to nested methods. Instead of that, partially moves responsibility to create IXamlType from CecilTypeSystem to CecilTypeResolver, which internally abstracts on top of both CecilTypeSystem and GenericTypeResolver with GenericParameterResolver. 
Every IXamlType instance now goes through generics transformation, except cases where it's explicitly disable (IXamlType.GenericTypeDefinition). 

This and also multiple type caching layers were reduced to one in CecilTypeResolver. Which is greatly simplified using TypeReferenceEqualityComparer. 

Fixes https://github.com/AvaloniaUI/Avalonia/issues/10485
Fixes https://github.com/AvaloniaUI/Avalonia/issues/9090
Probably https://github.com/AvaloniaUI/Avalonia/issues/8684 (didn't test)
Probably https://github.com/AvaloniaUI/Avalonia/issues/3395 (didn't test)

### CODE REVIEW

File: src/XamlX.IL.Cecil/CecilField.cs [hunk 1]
```diff
@@ -6,41 +6,49 @@ namespace XamlX.TypeSystem
...
+            private readonly CecilTypeResolver _typeResolver;
+            public FieldDefinition Field { get; }
 
-            public CecilField(CecilTypeSystem typeSystem, FieldDefinition def, TypeReference declaringType)
+            public CecilField(CecilTypeResolver typeResolver, FieldDefinition def)
             {
-                TypeSystem = typeSystem;
+                _typeResolver = typeResolver;
                 _def = def;
-                Field = new FieldReference(def.Name, def.FieldType, declaringType);
```
Comment (maxkatz6) [inline src/XamlX.IL.Cecil/CecilField.cs:19]: Is there any reason we have creating a copy of the FieldReference here from the definition?
Note, new type resolved doesn't mutate input references, so it's probably not needed anymore.

File: src/XamlX.IL.Cecil/CecilMethod.cs [hunk 2]
```diff
@@ -8,32 +8,26 @@ namespace XamlX.TypeSystem
...
             public MethodDefinition Definition { get; }
-            protected readonly TypeReference _declaringTypeReference;
 
-            public CecilMethodBase(CecilTypeSystem typeSystem, MethodReference method, TypeReference declaringType)
+            public CecilMethodBase(CecilTypeResolver typeResolver, MethodReference method)
             {
-                TypeSystem = typeSystem;
-
-                MethodReference MakeRef(bool transform)
+                MethodReference MakeIlRef()
```
Comment (maxkatz6) [inline src/XamlX.IL.Cecil/CecilMethod.cs]: MakeIlRef is only used to create a copy for IlReference. That's an old behavior which I didn't want to change. 
`Reference` property (non a IlReference) uses type resolver as the rest of types. Also `Reference` is used to resolve generics for return type and arguments.

File: src/XamlX.IL.Cecil/CecilTypeBuilder.cs [hunk 3]
```diff
@@ -11,11 +11,11 @@ partial class CecilTypeSystem
...
         {
-            protected TypeReference SelfReference;
-            public CecilTypeBuilder(CecilTypeSystem typeSystem, CecilAssembly assembly, TypeDefinition definition) 
-                : base(typeSystem, assembly, definition)
+            private CecilTypeResolver _builderTypeResolver;
+            public CecilTypeBuilder(CecilTypeResolver parentTypeResolver, CecilAssembly assembly, TypeDefinition definition) 
+                : base(parentTypeResolver, assembly, definition)
             {
-                SelfReference = definition;
+                _builderTypeResolver = parentTypeResolver.Nested(definition);
```
Comment (maxkatz6) [inline src/XamlX.IL.Cecil/CecilTypeBuilder.cs]: SelfReference type was replaced with a `_builderTypeResolver`. In this context, it has the same meaning, as all nested members will be resolved from it, like they did previously from SelfReference.

File: src/XamlX.IL.Cecil/CecilTypeSystem.cs [hunk 4]
```diff
@@ -105,60 +99,12 @@ public IXamlType FindType(string name, string assembly)
...
         public MethodReference GetMethodReference(IXamlConstructor t) => ((CecilConstructor)t).IlReference;
 
-        CecilAssembly FindAsm(AssemblyDefinition d)
+        internal CecilAssembly FindAsm(AssemblyDefinition d)
         {
             _assemblyDic.TryGetValue(d, out var asm);
             return asm;
         }
 
-        static string GetTypeReferenceKey(TypeReference reference)
```
Comment (maxkatz6) [inline src/XamlX.IL.Cecil/CecilTypeSystem.cs:114]: Replaced with TypeReferenceEqualityComparer.

File: src/XamlX.IL.Cecil/Comparers/CecilTypeComparisonMode.cs [hunk 5]
```diff
@@ -0,0 +1,11 @@
+﻿namespace XamlX.TypeSystem;
+
+public enum CecilTypeComparisonMode {
+    Exact,
```
Comment (maxkatz6) [inline src/XamlX.IL.Cecil/Comparers/CecilTypeComparisonMode.cs:4]: Exact is used in type caching, as it's a default value. We probably could go away with SignatureOnly for this purpose, but I didn't try.

File: src/XamlX.IL.Cecil/CecilEvent.cs [hunk 6]
```diff
@@ -1,33 +1,39 @@
...
-            public bool Equals(IXamlEventInfo other) => other is CecilEvent cp && cp.Event == Event;
+
             public string Name => Event.Name;
 
             private IXamlMethod _getter;
 
             public IXamlMethod Add => Event.AddMethod == null
                 ? null
-                : _getter ?? (_getter = TypeSystem.Resolve(Event.AddMethod, _declaringType));
+                : _getter ??= new CecilMethod(_typeResolver, Event.AddMethod);
```
Comment (kekekeks) [inline src/XamlX.IL.Cecil/CecilEvent.cs]: So no way to cache those now, right?

File: src/XamlX.IL.Cecil/Resolvers/GenericParameterResolver.cs [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+
+	internal static TypeReference ResolveParameterTypeIfNeeded(PropertyReference property, ParameterReference parameter)
+	{
+		var declaringGenericInstanceType = property.DeclaringType as GenericInstanceType;
+
+		if (declaringGenericInstanceType == null)
+			return parameter.ParameterType;
+
+		return ResolveIfNeeded(null, declaringGenericInstanceType, parameter.ParameterType);
+	}
```
Comment (maxkatz6) [inline src/XamlX.IL.Cecil/Resolvers/GenericParameterResolver.cs:55]: This method is pretty much the only addition from me here, as property related methods were missing in the original code.

Comment (maxkatz6) [reply src/XamlX.IL.Cecil/Resolvers/GenericParameterResolver.cs:55]: For the original code, see file headers.

File: src/XamlX.IL.Cecil/CecilEvent.cs (hunk 6 above)
Comment (kekekeks) [reply src/XamlX.IL.Cecil/CecilEvent.cs]: Also, how does it resolve generic arguments from declaring generic type instance now? IIRC EventDefinition is pointing to TypeDefinition metadata that lacks that information.

Comment (kekekeks) [reply src/XamlX.IL.Cecil/CecilEvent.cs]: Ah, it's moved to the resolver. I think we need CecilTypeResolver to be renamed to CecilTypeResolveContext or something, given that it's parent type/method specific.

File: src/XamlX.IL.Cecil/CecilField.cs (hunk 1 above)
Comment (kekekeks) [reply src/XamlX.IL.Cecil/CecilField.cs:19]: We were force-converting definition to a reference, probably for immutability, yes.

**Comment by maxkatz6** [comment]
Ah nice. Tests are failing, but only on Linux.

**Review by kekekeks — COMMENTED** [review]
Do we now support building and reflecting over open generics too?

**Comment by kekekeks** [comment]
>Ah nice. Tests are failing, but only on Linux.

Mono

File: src/XamlX.IL.Cecil/CecilEvent.cs (hunk 6 above)
Comment (maxkatz6) [reply src/XamlX.IL.Cecil/CecilEvent.cs]: Yes, "context" makes more sense.

Comment (maxkatz6) [reply src/XamlX.IL.Cecil/CecilEvent.cs]: > So no way to cache those now, right?

I don't think methods were cached previously? But yes, caching is hard.

File: src/XamlX.IL.Cecil/CecilField.cs (hunk 1 above)
Comment (maxkatz6) [reply src/XamlX.IL.Cecil/CecilField.cs:19]: Apparently, we needed FieldReference here, as passing FieldDefinition to the IL Emitter is a bad idea.

**Review by kekekeks — APPROVED** [review]

**Pull request merged**